tabgroups/init.lua

```diff
@@ -0,0 +1,840 @@
+    { "<Return>", "Switch to tabgroup or tab in tabgroup.", switch_tabgroup_or_tab },
+    { "n", "Create new tabgroup.", new_tabgroup },
+    { "r", "Rename tabgroup.", rename_tabgroup },
+    { "d", "Delete tabgroup.", function(w)
+        local row = w.menu:get()
+        w:set_mode()
+        if row and row._group then
+            local deleted = delete_tabgroup(w, row._group)
+            open_tabgroup_menu(w)
+            if not deleted then
+                w:notify("Can't remove last tabgroup")
+            end
+        end
+    end},
+    { "+", "Open list of tabs in tabgroup.", show_tabgroup_content },
+    { "-", "Hide list of tabs in tabgroup.", show_tabgroups },
+    { "m", "Move tab to different tabgroup.", move_tab_to_tabgroup_menu },
+    { "1", "Switch to first tab or tabgroup.", switch_tabgroup_or_tab, { count = 1 } },
+    { "2", "Switch to second tab or tabgroup.", switch_tabgroup_or_tab, { count = 2 }  },
+    { "3", "Switch to third tab or tabgroup.", switch_tabgroup_or_tab, { count = 3 }  },
+    { "4", "Switch to forth tab or tabgroup.", switch_tabgroup_or_tab, { count = 4 }  },
+    { "5", "Switch to fifth tab or tabgroup.", switch_tabgroup_or_tab, { count = 5 }  },
+    { "6", "Switch to sixth tab or tabgroup.", switch_tabgroup_or_tab, { count = 6 }  },
+    { "7", "Switch to seventh tab or tabgroup.", switch_tabgroup_or_tab, { count = 7 }  },
+    { "8", "Switch to eighth tab or tabgroup.", switch_tabgroup_or_tab, { count = 8 }  },
+    { "9", "Switch to ninth tab or tabgroup.", switch_tabgroup_or_tab, { count = 9 }  },
+}, menu_binds))
+
+add_binds("normal", {
+    { "x", "Open tabgroup menu.", open_tabgroup_menu },
+    { "X", "Show tabs in active tabgroup.", function (w) open_tabgroup_menu(w, w2groups[w].active)  end },
+})
+
+settings.register_settings({
+    ["tabgroups.sort_groups_by"] = {
+        type = "string",
+        default = "name asc",
+        validator = function(v)
+            local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
+            if not field then return false end
+            field, order = field:lower(), order:lower()
+            field = ({name = true, atime = true, mtime = true, ctime = true, hits = true})[field]
+            order = order == "asc" or order == "desc"
+            return field and order
+        end,
+        desc = [=[ Sort order of groups in tabgroups menu.
+
+Must be in the form "_field_ _order_", where _field_ is one of:
+
+- `name`: group name
+- `ctime`: time of group creation
+- `mtime`: time of group modification (updated when adding or removing new tabs)
+- `atime`: time of group accessing (updated when switching to this group)
+- `hits`: number of times this group has been switched to
+
+and _order_ is
+
+- `asc`: sort in ascending order
+- `desc`: sort in descending order
+]=]
+    },
+    ["tabgroups.sort_tabs_by"] = {
+        type = "string",
+        default = "title asc",
+        validator = function(v)
```

Comment:
This validation function can be made much shorter, the same way as the `tabgroups.sort_groups_by` was.

Reply:
this is diff for latest changes:
diff --git a/tabgroups/init.lua b/tabgroups/init.lua
index 38a4aec..38ff341 100644
--- a/tabgroups/init.lua
+++ b/tabgroups/init.lua
@@ -300,7 +300,8 @@ session.add_signal("restore", function (state)
                         open_new_tab_in_tabgroup(w, gn, stat.uri,  {
                             switch = g.active == i,
                             session_state = stat.session_state,
-                            session_restore = true })
+                            session_restore = true
+                        })
                     end
                     local group = w2groups[w].groups[gn]
                     group.tabs[i]  = {
@@ -381,7 +382,8 @@ local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
             mtime = tab.mtime,
             ctime = tab.ctime,
             title = v.title,
-            n = i })
+            n = i
+        })
     end
     if field and order then
         table.sort(_tmp, function(a, b) return _sort_by_field(field, order, a, b) end)
@@ -394,13 +396,15 @@ local function _build_tabgroup_menu_grouptabs(w, group_name, field, order)
                 "<b>"..((i < 10 and i..' - ') or '')..group_name.."</b>",
                 "<b>"..lousy.util.escape(title).."</b>",
                 lousy.util.escape(v.v.uri),
-                _group = group_name, _tab = v.v})
+                _group = group_name, _tab = v.v
+            })
         else
             table.insert(rows, {
                 ((i < 10 and i..' - ') or '')..lousy.util.escape(group_name),
                 lousy.util.escape(title),
                 lousy.util.escape(v.v.uri),
-                _group = group_name, _tab = v.v})
+                _group = group_name, _tab = v.v
+            })
         end
     end
     return rows
@@ -427,12 +431,14 @@ local function _build_tabgroup_menu_grouplist(w, field, order)
             table.insert(rows, {
                 "<b>"..((i < 10 and i..' - ') or '')..lousy.util.escape(g.name).."</b>",
                 w2groups[w].groups[g.name]._notebook:count(),
-                _group = g.name})
+                _group = g.name
+            })
         else
             table.insert(rows, {
                 ((i < 10 and i..' - ') or '')..lousy.util.escape(g.name),
                 w2groups[w].groups[g.name]._notebook:count(),
-                _group = g.name})
+                _group = g.name
+            })
         end
     end
     return rows
@@ -758,7 +764,7 @@ settings.register_settings({
         default = "name asc",
         validator = function(v)
             local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
-            if not field then return false end
+            if not field or not order then return false end
             field, order = field:lower(), order:lower()
             field = ({name = true, atime = true, mtime = true, ctime = true, hits = true})[field]
             order = order == "asc" or order == "desc"
@@ -785,27 +791,10 @@ and _order_ is
         default = "title asc",
         validator = function(v)
             local field, order = string.match(v, "^%s*(%w+)%s+(%w+)")
-            if field then
-                field = string.lower(field)
-                if field ~= 'n' and
-                    field ~= 'title' and
-                    field ~= 'atime' and
-                    field ~= 'mtime' and
-                    field ~= 'ctime' and
-                    field ~= 'hits' then
-                    return false
-                end
-            else
-                return false
-            end
-            if order then
-                order = string.lower(order)
-                if order ~='asc' and order ~= 'desc' then
-                    return false
-                end
-            else
-                return false
-            end
-            return true
+            if not field or not order then return false end
+            field, order = field:lower(), order:lower()
+            field = ({n = true, title = true, atime = true, mtime = true, ctime = true, hits = true})[field]
+            order = order == "asc" or order == "desc"
+            return field and order
         end,
         desc = [=[ Sort order of tabs in tabgroups menu.